rocAL/rocAL/source/mxnet_meta_data_reader.cpp

```diff
@@ -0,0 +1,185 @@
+    }
+}
+
+void MXNetMetaDataReader::read_all(const std::string &_path)
+{
+    std::string _rec_file, _idx_file;
+    if ((_src_dir = opendir (_path.c_str())) == nullptr)
+        THROW("MXNetMetaDataReader ERROR: Failed opening the directory at " + _path);
+
+    while((_entity = readdir (_src_dir)) != nullptr)
+    {
+        std::string file_name = _path + "/" + _entity->d_name;
+        filesys::path pathObj(file_name);
+        if(filesys::exists(pathObj) && filesys::is_regular_file(pathObj))
+        {
+            auto file_extension_idx = file_name.find_last_of(".");
+            if (file_extension_idx  != std::string::npos)
+            {
+                std::string file_extension = file_name.substr(file_extension_idx+1);
+                if (file_extension == "rec")
+                    _rec_file = file_name;
+                else if(file_extension == "idx")
+                    _idx_file = file_name;
+                else
+                    continue;
+            }
+        }
+    }
+    closedir(_src_dir);
+    uint rec_size;
+    _file_contents.open(_rec_file);
+    if (!_file_contents)
+        THROW("MXNetMetaDataReader ERROR: Failed opening the file " + _rec_file);
+    _file_contents.seekg(0, ifstream::end);
+    rec_size = _file_contents.tellg();
+    _file_contents.seekg(0, ifstream::beg);
+
+    ifstream index_file(_idx_file);
+    if(!index_file)
+        THROW("MXNetMetaDataReader ERROR: Could not open RecordIO index file. Provided path: " + _idx_file);
+
+    std::vector<size_t> _index_list;
+    size_t _index, _offset;
+    while (index_file >> _index >> _offset)
+        _index_list.push_back(_offset);
+    if(_index_list.empty())
+        THROW("MXNetMetaDataReader ERROR: RecordIO index file doesn't contain any indices. Provided path: " + _idx_file);
+    _index_list.push_back(rec_size);
+    std::sort(_index_list.begin(), _index_list.end());
+    for (size_t i = 0; i < _index_list.size() - 1; ++i)
+        _indices.emplace_back(_index_list[i], _index_list[i + 1] - _index_list[i]);
+    read_images();
+}
+
+void MXNetMetaDataReader::release(std::string _image_name)
+{
+    if(!exists(_image_name))
+    {
+        WRN("ERROR: Given not present in the map" + _image_name);
+        return;
+    }
+    _map_content.erase(_image_name);
+}
+
+void MXNetMetaDataReader::release() {
+    _map_content.clear();
+}
+
+void MXNetMetaDataReader::read_images()
+{
+    for(int current_index = 0; current_index < (int)_indices.size(); current_index++ )
+    {
+        uint32_t _magic, _length_flag;
+        int64_t _seek_pos, _data_size_to_read;
+        std::tie(_seek_pos, _data_size_to_read) = _indices[current_index];
+        _file_contents.seekg(_seek_pos, ifstream::beg);
+        uint8_t* _data = new uint8_t[_data_size_to_read];
+        uint8_t* _data_ptr = _data;
+        _file_contents.read((char *)_data_ptr, _data_size_to_read);
+        memcpy(&_magic, _data_ptr, sizeof(_magic));
```

Comment:
you can avoid this memcpy by directly reading data  (_magic = *((uint32_t *) _data_ptr);